### ISSUE TITLE
`home.keyboard.options` lost on keyboard unplug/plug

### ISSUE THREAD

**Issue by sir4ur0n** [nix-community/home-manager#1638]
Hello and thank you for Home Manager, it's really awesome :bow: 

### Issue description

Whenever I unplug and plug back my keyboard (I use the same keyboard on 2 computers, so it's common), the keyboard configuration (in particular the `nocaps` bit) gets lost.

The - hopefully - relevant part of my `home.nix`:
```nix
  home.keyboard = {
    layout = "us,fr";
    options = [ "grp:alt_space_toggle" "ctrl:nocaps" ];
  };
```

Right after unplugging my keyboard:
```
$ setxkbmap -print -verbose 10
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /nix/store/cm6z25y2nwcv5s8cnh60bnmakmr5r61f-setxkbmap-1.3.2/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc104
layout:     us
options:    terminate:ctrl_alt_bksp
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+us+inet(evdev)+terminate(ctrl_alt_bksp)
geometry:   pc(pc104)
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+us+inet(evdev)+terminate(ctrl_alt_bksp)"	};
	xkb_geometry  { include "pc(pc104)"	};
};
```
:warning: In particular, the Caps Lock key is back to turning all caps on, rather than behaving as Ctrl key.

Now run `systemctl --user restart setxkbmap.service`.

:heavy_check_mark: Caps Lock now correctly behaves as Ctrl again.

```
$ setxkbmap -print -verbose 10
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /nix/store/cm6z25y2nwcv5s8cnh60bnmakmr5r61f-setxkbmap-1.3.2/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc104
layout:     us,fr
options:    grp:alt_space_toggle,ctrl:nocaps,terminate:ctrl_alt_bksp
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+us+fr:2+inet(evdev)+group(alt_space_toggle)+ctrl(nocaps)+terminate(ctrl_alt_bksp)
geometry:   pc(pc104)
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+us+fr:2+inet(evdev)+group(alt_space_toggle)+ctrl(nocaps)+terminate(ctrl_alt_bksp)"	};
	xkb_geometry  { include "pc(pc104)"	};
};
```

TL;DR:
* Keyboard options lost on unplug/plug
* Workaround is running `systemctl --user restart setxkbmap.service` after plugging the USB keyboard

#### Maintainer CC

Based on https://github.com/nix-community/home-manager/blob/master/modules/home-environment.nix#L168
@rycee 

#### Technical details

```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.72, NixOS, 21.03pre249162.1dc37370c48 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.8`
 - channels(sir4ur0n): `"nixpkgs-unstable-21.03pre252431.4f3475b113c, home-manager"`
 - channels(root): `"nixos-21.03pre249162.1dc37370c48"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by berbiche** [comment]
If the xkb config is not preserved for an input then the best option is to listen for udev events and execute the script when a keyboard is connected.
Most Desktop Environment (GNOME, KDE, XFCE, etc.) allow configuring these settings and I believe they listen for udev events.

**Comment by Sir4ur0n** [comment]
Hi @berbiche , I'm not sure I understood your point :sweat_smile: 
Is that something that could be taken care of by Home Manager? Or is it impossible because Home Manager can't modify udev rules?

Maybe this problem and a solution/link to solution(s) should be added to the `home.keyboard.options` option?

**Comment by berbiche** [comment]
> Is that something that could be taken care of by Home Manager? Or is it impossible because Home Manager can't modify udev rules?
> 
> Maybe this problem and a solution/link to solution(s) should be added to the `home.keyboard.options` option?

I don't think it's possible for Home Manager to manage udev rules.
- https://wiki.archlinux.org/index.php/udev
- man udev(7)

Paths from which rules are read are all managed outside of a user's home.

**Comment by rycee** [comment]
The plugging/unplugging events are probably available to the user, you can try running `udevadm monitor` as a non-root user and see if it prints a bunch of information.

That said, I don't want HM to do anything like this. It should use some 3rd party daemon that provides this functionality. HM can generate a configuration and systemd service file for it.

**Comment by ronthecookie** [comment]
> That said, I don't want HM to do anything like this. 
It might be worthwhile to look into how `localectl` from systemd does this as I couldn't reproduce this issue when using that.

**Comment by rycee** [comment]
Interesting, if localectl does better than setxkbmap, then it would be best to simply switch over to using localectl.

**Comment by ronthecookie** [comment]
It seems it blows up on NixOS:
```
ron@cookiemonster ~ -> localectl set-x11-keymap us,il
Failed to set keymap: Changing system settings via systemd is not supported on NixOS.
```

**Comment by rycee** [comment]
Bummer, hopefully it will be possible in the future.

**Comment by ronthecookie** [comment]
I looked around and it seems they just escalate privileges and [write](https://github.com/systemd/systemd/blob/main/src/locale/keymap-util.c#L422) to the Xorg config.

**Comment by akho** [comment]
There is a daemon called [xplugd](https://github.com/troglobit/xplugd) that handles this. I don't see it being in Nixpkgs, though.

**Comment by ckiee** [comment]
@akho That looks like it might fix another one of my problems; I might package it soon.

**Comment by akho** [comment]
@ckiee I've packaged and submitted a pull request at NixOS/nixpkgs#131368.

**Comment by akho** [comment]
xplugd is now packaged in nixpkgs master.

I'm not sure how (and whether) to write a module for it. I basically have this in `home.nix`:

    systemd.user.services.xplugd = {
      Unit = {
        Description = "xplugd";
        PartOf = [ "graphical-session.target" ];
      };
      Install.WantedBy = [ "graphical-session.target" ];
      Service.ExecStart = "${pkgsUnstable.xplugd}/bin/xplugd -n";
    };
    xdg.configFile."xplugrc" = {
      executable = true;
      text = ''#!/bin/sh
        case "$1,$3,$4" in
            keyboard,connected,*)
                systemctl --user restart setxkbmap.service
                ;;
        esac
      '';
    };

I'm not sure if the setxkbmap restart should be there by default (if kbd options are set), and whether the script should be just an option with a bit of text, or something more nixified (the case statement could probably be an attribute hierarchy?).
